### PR TITLE
[Windows] Fixes to Hermes from source on Windows

### PR DESCRIPTION
**Pull request by mganandraj**
This change helps in building React native with Hermes from source on Windows boxes.
    
    Specifically, this change adds
    1. Explicitly use ninja with CMake to build the Hermes compiler. VS generators works, but as VS is multi-config, it requires the build flavor to be explicitly specified.
    And the build flavor of the Hermes compiler and runtime builds should match which will require us to build multiple flavors of compiler for debug and ship. And it is tricky to map the compiler and runtime builds to match flavor.
    Using Ninja generator avoids the complication.
    2. CMake shipped with Android SDK can't find the python installation on Windows with FindPython, but it works fine with the older FindPythonInterp. The switch was made very recently in hermes repo (https://github.com/facebook/hermes/commit/5244f819b2f3949ca94a3a1bf75d54a8ed59d94a)
    This change adds a hook for the developer to hint FindPython module to find Python interpreter on Windows
## Summary

Fixes to build Hermes from source on Windows by ensuring the python interpret is resolvable by CMake, and avoiding the need to match the hermes compiler and runtime build flavors..

## Changelog

This change helps in building React native with Hermes from source on Windows boxes.
    
    Specifically, this change adds
    1. Explicitly use ninja with CMake to build the hermes compiler. VS generators works, but as VS is multi-config, it requires the build flavor to be explicitly specified.
    And the the build flavor of the hermes compiler and runtime builds should match which will require us to build multiple flavors of compiler for debug and ship. And it is tricky to map the compiler and runtime builds to match flavor.
    Using Ninja generator avoids the complication.
    2. CMake shipped with Android SDK can't find the python installation on Windows with FindPython, but it works fine with the older FindPythonInterp. The switch was made very recently in hermes repo (https://github.com/facebook/hermes/commit/5244f819b2f3949ca94a3a1bf75d54a8ed59d94a)
    This change adds a hook for the developer to hint FindPython module to find Python interpreter on Windows
[CATEGORY] [TYPE] - Message

## Test Plan

Verified building and running Hermes on Windows

### CODE REVIEW

**Comment by mganandraj** [comment]
@cortinico @kelset @neildhar These are a couple of changes to build Hermes on Windows ..

**Comment by mganandraj** [comment]
> that's something that we need only on 0.69 and main right?

Yes.. these changes pertains to building Hermes from source, which is in 0.69/main.. 

Also, we should fix the documentation for windows users.. For instance, the user would require an MSVC installation (or another windows tool chain that cmake/ninja is compatible with, llvm or mingw), add ninja from ndk to path etc. How should I go about it ?

**Comment by kelset** [comment]
> Also, we should fix the documentation for windows users.. For instance, the user would require an MSVC installation (or another windows tool chain that cmake/ninja is compatible with, llvm or mingw), add ninja from ndk to path etc. How should I go about it ?

if the documentation is on the https://reactnative.dev/ website, the repo is here: https://github.com/facebook/react-native-website, it should be a matter of doing changes in one markdown file since it'd have to be changed for >=0.69 and that version has not been cut

**Comment by cortinico** [comment]
Thanks for sending this over.

> Context: CMake in Android SDK can't find the python installation on Windows

Do you have some reference on why this is happening or what's the error message?

**Comment by cortinico** [comment]
> For instance, the user would require an MSVC installation (or another windows tool chain that cmake/ninja is compatible with, llvm or mingw), add ninja from ndk to path etc. How should I go about it ?

Following up here: I'll be writing a page on the website about this change. I can share it with as soon as it's ready so we can include those changes

**Comment by mganandraj** [comment]
> 



> 

What i'm seeing is related to the issue described here: https://gitlab.kitware.com/cmake/cmake/-/issues/19024
Essentially, CMake tries to make sure that the resolved python has the right bit-ness.
But what i'm seeing is a bit subtler than the above issue, as we query only for Interpreter (find_package(Python COMPONENTS Interpreter))

When using 32-bit VS development console, CMake can't find the global python interpreter installed through Windows Store when using Ninja generator. CMake still can find the global python interpreter when using VS generator with '-A x64'.

When using 64-bit VS development console, CMake can find the global Python interpreter. But setting up 64-bit powershell VS development console is a bit involved.

In short, Windows has this complication and it's highly likely that the users will hit this depending on their particular python  installation and dev processes. It would be useful to provide this hook to override as this works more deterministically. Another plus is that this hooks will allow developers to use the python interpreter available inside the ndk !

**Comment by mganandraj** [comment]
> 

What i'm seeing is related to the issue described here: https://gitlab.kitware.com/cmake/cmake/-/issues/19024
Essentially, CMake tries to make sure that the resolved python has the right bit-ness.
But what i'm seeing is a bit subtler than the above issue, as we query only for Interpreter (find_package(Python COMPONENTS Interpreter))

When using 32-bit VS development console, CMake can't find the global python interpreter installed through Windows Store when using Ninja generator. CMake still can find the global python interpreter when using VS generator with '-A x64'.

When using 64-bit VS development console, CMake can find the global Python interpreter. But setting up 64-bit powershell VS development console is a bit involved.

In short, Windows has this complication and it's highly likely that the users will hit this depending on their particular python installation and dev processes. It would be useful to provide this hook to override as this works more deterministically. Another plus is that this hooks will allow developers to use the python interpreter available inside the ndk !

**Comment by neildhar** [comment]
I'm not at all familiar with how the new FindPython module interacts with finding python on Windows, so I defer to Windows experts here.

However, I'm not sure why we need to add a dependency on ninja here, as opposed to using the default build system on the host.

> build flavor of the Hermes compiler and runtime builds should match

This isn't really true. You can just do a single build of the compiler here and use it for both the debug and release builds of Hermes. The bytecode does not change depending on build flavour. It's worth switching this to always build in release mode though, so that compilation is fast.

**Comment by mganandraj** [comment]
> I'm not at all familiar with how the new FindPython module interacts with finding python on Windows, so I defer to Windows experts here.
> 
> However, I'm not sure why we need to add a dependency on ninja here, as opposed to using the default build system on the host.
> 
> > build flavor of the Hermes compiler and runtime builds should match
> 
> This isn't really true. You can just do a single build of the compiler here and use it for both the debug and release builds of Hermes. The bytecode does not change depending on build flavour. It's worth switching this to always build in release mode though, so that compilation is fast.

@neildhar This is what I realized while looking into this. When building with VS generator, the export script from hermesc build  (ImportHermesc.cmake) contains macros to lookup all build configuration, irrespective which flavor is actually built. 

    # Import target "native-hermesc" for configuration "Debug"
    set_property(TARGET native-hermesc APPEND PROPERTY IMPORTED_CONFIGURATIONS DEBUG)
    set_target_properties(native-hermesc PROPERTIES IMPORTED_LOCATION_DEBUG "D:/github/react-   native/ReactAndroid/hermes-engine/build/hermes/bin/Debug/hermesc.exe" )
    
    # Import target "native-hermesc" for configuration "Release"
    set_property(TARGET native-hermesc APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
    set_target_properties(native-hermesc PROPERTIES IMPORTED_LOCATION_RELEASE "D:/github/react-native/ReactAndroid/hermes-engine/build/hermes/bin/Release/hermesc.exe"  )
    
    # Import target "native-hermesc" for configuration "MinSizeRel"
    set_property(TARGET native-hermesc APPEND PROPERTY IMPORTED_CONFIGURATIONS MINSIZEREL)
    set_target_properties(native-hermesc PROPERTIES IMPORTED_LOCATION_MINSIZEREL "D:/github/react-native/ReactAndroid/hermes-engine/build/hermes/bin/MinSizeRel/hermesc.exe"  )
    
    # Import target "native-hermesc" for configuration "RelWithDebInfo"
    set_property(TARGET native-hermesc APPEND PROPERTY IMPORTED_CONFIGURATIONS RELWITHDEBINFO)
    set_target_properties(native-hermesc PROPERTIES  IMPORTED_LOCATION_RELWITHDEBINFO "D:/github/react-native/ReactAndroid/hermes-engine/build/hermes/bin/RelWithDebInfo/hermesc.exe"  )

While, when using ninja, the export script contains only paths to the artefacts which was actually built. For e.g. by default the debug flavor is built and the export script contains only,

    # Import target "native-hermesc" for configuration "Debug"
    set_property(TARGET native-hermesc APPEND PROPERTY IMPORTED_CONFIGURATIONS DEBUG)
    set_target_properties(native-hermesc PROPERTIES IMPORTED_LOCATION_DEBUG "D:/testapps/abcdefghijk/react-native/ReactAndroid/hermes-engine/build/hermes/bin/hermesc.exe")

Later when building the runtime library by consuming ImportHermesC.cmake, CMake looks for the hermesc.exe matching the build flavor. For e.g. when building the release flavor runtime, CMake tries to locate IMPORTED_LOCATION_RELEASE, and when building the minsizerel flavor runtime (RN release build), CMake tries to locate IMPORTED_LOCATION_MINSIZEREL with VS generator. I haven't found a way yet to override the default resolution to explicitly provide the flavor of the imported artefact.

While with ninja, all flavor of the runtime builds are able to link against the default flavor compiler build.

**Comment by mganandraj** [comment]
> 

@cortinico @neildhar 

Ah .. I played around a bit more .. And this is what I observed,
1. When using CMake in Android SDK to generate the build, the default generator is actually not "Visual Studio NN" as I assumed earlier, but is "NMake Makefiles". I guess, the confusion started as I was building Hermes standalone initially using global CMake (installed by VS) before switching to building from the RN gradle build which hardcodes to Android SDK's CMake.  Unlike VS/MSBuild, "NMake Makefiles" is a single config build driver, hence it works similar to Ninja. Hence, the first part of the PR which switches to Ninja generator is not required unless we want to allow developers to use alternate CMake.

2. Somehow the FindPython module can't find the python interpreter installed from the Store. But FindPython can find the interpreter executable explicitly added into path. As the later works, I think we can wait for the feedbacks to see whether the hook is required., hence the second part is also not absolutely required.

Hence, I',m closing this PR for now.

Sorry for the confusion 🙂

**Comment by cortinico** [comment]
Thanks for the clarification and thanks for looking into this 👌
